### PR TITLE
v1alpha4 -> v1beta1 clusterctl upgrade test

### LINKED ISSUE #1809: Add e2e upgrade tests for v1alpha4 -> v1beta1 #1727

### PR DESCRIPTION
**Pull request by shysank**
**What type of PR is this?**
/kind other

**What this PR does / why we need it**:
Adds v1alpha4 -> v1beta1 upgrade test
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1809 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -43,11 +43,13 @@ require (
 	k8s.io/klog/v2 v2.9.0
 	k8s.io/kubectl v0.22.2
 	k8s.io/utils v0.0.0-20210930125809-cb0fa318a74b
-	sigs.k8s.io/cluster-api v1.0.0
-	sigs.k8s.io/cluster-api/test v1.0.0
-	sigs.k8s.io/controller-runtime v0.10.2
+	sigs.k8s.io/cluster-api v0.0.0-20211028170527-10d89ceca938
```
Comment (CecileRobertMichon) [inline go.mod:46]: /hold

until this is in a CAPI release

Comment (shysank) [reply go.mod:46]: backport pr https://github.com/kubernetes-sigs/cluster-api/pull/5541

**Comment by CecileRobertMichon** [comment]
the pull CAPI job has gotten really long recently with all the new test specs + clusterctl upgrade. Wondering if we should breakout api upgrade into its own job so it runs in parallel, or if there is any way of increasing parallelism on the existing jobs. wdyt?

**Comment by shysank** [comment]
> the pull CAPI job has gotten really long recently with all the new test specs + clusterctl upgrade. Wondering if we should breakout api upgrade into its own job so it runs in parallel, or if there is any way of increasing parallelism on the existing jobs. wdyt?

We could make it parallel by using ginkgo's parallelism, don't think we are doing it now. I see we set `GINKGO_NODES` but I think we should also add `-p` to enable parallelism. But, I'm not sure how the environment variables would work in that case. For the longer term, we should try to make tests not rely on any global state, and then enable parallelism. But, the easier short term fix would be to just have another job, as you've pointed out. I'd even go as far to split them in half to save more time. It probably won't make sense to look at something like `capi_e2e_part1` and `capi_e2e_part2`. Maybe we can overcome that by having some kind of description that says what tests are being run.

**Comment by CecileRobertMichon** [comment]
I'm pretty sure we do use parallelism, if you look at a recent run eg. https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/1810/pull-cluster-api-provider-azure-capi-e2e/1454146115451490304, the total duration was 2h48, with 14 tests that ran at around ~30 minutes each (rough average), if they were running serially that would have taken 7 hours. The fact that we're running them on 3 Ginkgo nodes makes sense since that's roughly 3x the actual duration.

**Comment by shysank** [comment]
> I'm pretty sure we do use parallelism, if you look at a recent run eg. https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/1810/pull-cluster-api-provider-azure-capi-e2e/1454146115451490304, the total duration was 2h48, with 14 tests that ran at around ~30 minutes each (rough average), if they were running serially that would have taken 7 hours. The fact that we're running them on 3 Ginkgo nodes makes sense since that's roughly 3x the actual duration.

yeah just saw that :)  updated my comment

**Comment by CecileRobertMichon** [comment]
race condition :)

let's try to bump the number of nodes and see if it does anything good/bad? we might be limited on the prow side of things

**Comment by shysank** [comment]
KCP upgrade in HA cluster failed, unrelated to this pr, `v1alpha4 -> v1beta1` upgrade test passed though. Trying again.
/test pull-cluster-api-provider-azure-capi-e2e

**Comment by shysank** [comment]
/hold cancel
@CecileRobertMichon This pr is ready now with capi upgraded to v1.0.1, ptal whenever you get a chance.

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

> Test started yesterday at 9:35 PM passed after 2h31m1s.

did you get anywhere with #1813? 2.5 hours is really long for a PR test

**Comment by shysank** [comment]
> did you get anywhere with #1816? 2.5 hours is really long for a PR test

I tried a bunch of things. There is some weird things happening with ginkgo nodes which I can't find the root cause yet. I'll keep trying whenever I get time, but meanwhile, wdyt think about splitting the jobs as you had suggested earlier?

**Comment by CecileRobertMichon** [comment]
+1 to splitting the clusterctl upgrade tests into a separate job

**Comment by CecileRobertMichon** [comment]
We could also remove the KCP upgrade specs from the CAPI job and run them as part of k8s upgrade jobs once #1735 merges, see https://github.com/kubernetes-sigs/cluster-api/issues/4896#issuecomment-970259100

File: test/e2e/capi_test.go [hunk 2]
```diff
@@ -257,5 +258,54 @@ var _ = Describe("Running the Cluster API E2E tests", func() {
 				}
 			})
 		})
+
+		Context("upgrade from v1alpha4 to v1beta1, and scale workload clusters created in v1alpha4", func() {
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply test/e2e/capi_test.go:262]: same for the other clusterctl upgrade spec that's already merged

[the remark this answers is not in the document]
Comment (shysank) [reply test/e2e/capi_test.go]: done, I've created a context `API Version Upgrade` that embeds the 2 upgrade tests. Is that okay? Happy to switch it to the way you've suggested if that's the convention.

[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply test/e2e/capi_test.go]: Yeah that should work

[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply test/e2e/capi_test.go]: are you going to open a PR to add it to Ginkgo skip and add a separate job?

[the remark this answers is not in the document]
Comment (shysank) [reply test/e2e/capi_test.go]: ptal at https://github.com/kubernetes/test-infra/pull/24392 whenever you get a chance

**Comment by CecileRobertMichon** [comment]
/lgtm
/approve

KCP upgrade spec failed but the expected tests ran

**Comment by CecileRobertMichon** [comment]
@shysank now that this merged, can we add it to the other beta release branch and as a periodic job?

**Comment by shysank** [comment]
> @shysank now that this merged, can we add it to the other beta release branch and as a periodic job?

do you mean backporting this pr to release-1.x, and add a job in `cluster-api-provider-azure-periodics-release-v1beta1` ?

**Comment by CecileRobertMichon** [comment]
we should add least run it periodically on the main branch, in `cluster-api-provider-azure-periodics-release-main`. 

Separately, we might also want to backport this to release-1.0 so we can add an apiversion test for both `cluster-api-provider-azure-periodics-release-v1beta1` and `cluster-api-provider-azure-pr-release-v1beta1` (the latter would be run on backport PRs). Would backporting this PR be feasible, including capi 1.0.1 bump?

**Comment by shysank** [comment]
> we should add least run it periodically on the main branch, in `cluster-api-provider-azure-periodics-release-main`.

+1
> Separately, we might also want to backport this to release-1.0 so we can add an apiversion test for both `cluster-api-provider-azure-periodics-release-v1beta1` and `cluster-api-provider-azure-pr-release-v1beta1` (the latter would be run on backport PRs). Would backporting this PR be feasible, including capi 1.0.1 bump?

yeah it should be possible to backport

**Comment by shysank** [comment]
ah deleted the comment by mistake :( the bot works though

**Pull request merged**